### ISSUE TITLE
Ability to indicate that a single type usage (in null-unmarked context) is known non-nullable [working decision: yes]

### ISSUE THREAD

**Issue by cpovirk** [jspecify/jspecify#229]
\[I edited this significantly after posting it to fit better with #230. My fault: kevinb9n had proposed that organizational scheme, but I'd forgotten.\]

This issue is one of _multiple_ issues that is taking over for #4, which is very long and which grew to cover multiple questions.

The new _overarching_ issue is #230. That covers whether a `@NonNull` annotation will exist.

_This_ issue is about one specific use case that motivates such an annotation -- specifically, the case of wanting to "sprinkle" annotations into otherwise unannotated code. Such "sprinkling" is already possible with `@Nullable`, but there's not currently an equivalent for marking types as _non_-nullable: If you want to mark a single type usage as non-null, the only way to do that is to annotate the whole method ([_probably_ the method](https://github.com/jspecify/jspecify/issues/43)) with `@NullMarked` -- which you can't do until you've added any necessary `@Nullable` annotations, too.

This support seems especially likely to be useful for declaring that particular _parameters_ are non-null. I say that because most nullness checkers effectively treat _return types_ as non-null by default.

With this ability, library owners can tactically annotate particular APIs known to be problematic. Also, they can use automated tools to add nullness information to _some_ parameters of a method even if those tools can't definitively prove the nullness of _all_ parameters.

**Comment by kevinb9n** [comment]
Meta-points about compatibility:

* With no `@NonNull` in the artifact, we believe we could still add it compatibly later. If we're overlooking an odd quirk -- like if adding it forces other small shifts in other decisions we'd committed, or something -- that would raise the urgency here.

* Likewise we also believe that adding it later should not prompt any released libraries to feel pressured to make incompatible changes in response.

**Comment by cpovirk** [comment]
For the record, I've edited the initial post here significantly to fit better with #230. For anyone reading over email, I'll re-post it here but then delete it, since it will be a duplicate of the newly revised initial post:

---

This issue is one of _multiple_ issues that is taking over for #4, which is very long and which grew to cover multiple questions.

The new _overarching_ issue is #230. That covers whether a `@NonNull` annotation will exist.

_This_ issue is about one specific use case that motivates such an annotation -- specifically, the case of wanting to "sprinkle" annotations into otherwise unannotated code. Such "sprinkling" is already possible with `@Nullable`, but there's not currently an equivalent for marking types as _non_-nullable: If you want to mark a single type usage as non-null, the only way to do that is to annotate the whole method ([_probably_ the method](https://github.com/jspecify/jspecify/issues/43)) with `@NullMarked` -- which you can't do until you've added any necessary `@Nullable` annotations, too.

This support seems especially likely to be useful for declaring that particular _parameters_ are non-null. I say that because most nullness checkers effectively treat _return types_ as non-null by default.

With this ability, library owners can tactically annotate particular APIs known to be problematic. Also, they can use automated tools to add nullness information to _some_ parameters of a method even if those tools can't definitively prove the nullness of _all_ parameters.

**Comment by cpovirk** [comment]
Here's a small point about a way you could try to "sprinkle" non-nullness _without_ `@NonNull`. It turns out to be a bad idea.

If you want to mark one type usage as non-null, you could try to mark the whole method as `@NullMarked` but then add `@Nullable // TODO: Is this right?` to all the parameters that you're unsure about. But:

- The biggest problem: You'd have put `@Nullable` on the _return type_, too. And that is _bad_ for callers if you're wrong. So probably what you'd really do is never annotate _any_ parameter unless you could _also_ annotate the return type properly. Under that approach, you still can't arbitrarily annotate single type usages.
- A `@Nullable` annotation may mislead users into thinking that `null` really "makes sense" as a parameter (especially if they're viewing only tooltips or Javadoc, not source code).
- (It would also prevent a "strict-mode" nullness checker from detecting problems. But I doubt many people will run in "strict mode," especially when using unannotated libraries like the one we're talking about here :))
- A `@Nullable` annotation prevents nullness checker from checking the _implementation_ as well.
- This approach isn't even sufficient when parameterized types are involved: A non-`@NullMarked` method that accepts `Foo<Bar>` can be called with a `Foo` of nullable `Bar` or a `Foo` of non-nullable `Bar`. But once the method is `@NullMarked`, it is declared to accept only one or the other. You can't "just mark `Bar` as `@Nullable`" and assume that callers will be OK.

**Comment by kevinb9n** [comment]
Yep looks a poor workaround and our docs will have to dissuade users from it.

Even just as a basic design principle, we want the user's natural motivations to point them toward annotating _correctly_ as much as possible.

I'm mostly looking at this issue through this kind of lens

* There will be files the developer isn't ready to migrate yet for whatever reason
* There will be type usages in such a file that the developer knows to be non-nullable
* It's at least odd or a bit sad if there is no way to record that knowledge, and it must just be forgotten

Of course, "odd or a bit sad" isn't an argument. But if an automated tool can recognize some of these cases, and might very well progressively recognize more over time, oddorabitsadness multiplies. This "sprinkling" seems like it could usefully defray the effort of the full migration.

Separately... how regularly do we think users might encounter problems IRL such that adding one `@NonNull` into an unmigrated file would be at least a decent remedy?

**Comment by cpovirk** [comment]
> How regularly do we think users might encounter problems IRL such that adding one `@NonNull` into an unmigrated file would be at least a decent remedy?

This is one of those excellent cases in which we could probably produce useful data: Just as we have [tools to spray `@Nullable` annotations](https://github.com/google/error-prone/tree/master/core/src/main/java/com/google/errorprone/bugpatterns/nullness) and Kotlin/CF clients that we can measure the impact on, so too could we write tools to spray `@NonNull` annotations and then measure the results.

(Well, we couldn't spray _JSpecify_ `@NonNull` annotations yet, but ironically, we haven't been spraying JSpecify _`@Nullable`_ annotations yet, either :))

Now, given that we haven't gotten very far with spraying even `@Nullable`, I'm not sure how soon we'd look into `@NonNull`. But it's a thing we could do, given the time!

**Comment by kevinb9n** [comment]
Working decision: yes, we will have `@Nullable` in 0.3 and it will work in the 2-3 ways expected.

**Comment by netdpb** [comment]
**Current decision**: `@NonNull` can be used in the same locations as `@Nullable` for this purpose.

**Argument for changing**: None.

**Timing**: This must be decided before version 1.0 of the jar.

**Proposal for 1.0: Finalize the current decision.** If you agree, please add a thumbs-up emoji (👍) to this comment. If you disagree, please add a thumbs-down emoji (👎) to this comment and briefly explain your disagreement. Please only add a thumbs-down if you feel you can make a strong case why this decision will be materially worse for users or tool providers than an alternative.

**Comment by netdpb** [comment]
**Decision:** Yes. `@NonNull` can be used for this purpose.
